src/main/java/gov/usgs/earthquake/indexer/ReliableIndexerListener.java

```diff
@@ -0,0 +1,258 @@
+import gov.usgs.util.Config;
+
+/**
+ * ReliableIndexerListener listens for product changes by the indexer, then handles the new products independently in a background thread.
+ * 
+ * This class does little more than output logs for the products it has seen; it is designed to be extended.
+ * 
+ * Several useful methods are availble to be overridden or otherwise used:
+ * <ul>
+ * <li>onBeforeProcessThreadStart</li>
+ * <li>onProcessException</li>
+ * <li>getNextProducts</li>
+ * <li>processProducts</li>
+ * </ul>
+ * 
+ * This class accepts an index for querying in config:
+ * 
+ * <dl>
+ * <dt>index</dt>
+ * <dd>(Required) The index to use for product querying.</dd>
+ * </dl>
+ */
+
+public class ReliableIndexerListener extends DefaultIndexerListener implements IndexerListener, Runnable {
+
+  protected static final Logger LOGGER = Logger
+          .getLogger(ReliableIndexerListener.class.getName());
+
+  private static int PRODUCTS_PER_QUERY = 10;
+
+  private boolean stopThread = false;
+  private long lastIndexId = -1;
+  private Object syncObject = new Object();
+
+  private Thread processThread;
+  protected ProductIndex productIndex;
+
+  /**
+   * Sets up an object on start
+   * 
+   * @param config configuration
+   *
+   * @throws Exception if missing product index
+   */
+  public void configure(Config config) throws Exception {
+    super.configure(config);
+
+    //Getting indexer for queries
+    String indexName = config.getProperty(Indexer.INDEX_CONFIG_PROPERTY);
+    if (indexName != null) {
+      LOGGER.config("[" + getName() + "] loading ProductIndex '"
+        + indexName + "'");
+      productIndex = (ProductIndex) Config.getConfig().getObject(indexName);
+    }
+    if (productIndex == null) {
+      throw new ConfigurationException("[" + getName()
+          + "] ProductIndex is required");
+    }
+  }
+
+  /** 
+   * Wakes thread when indexer makes changes
+   * 
+   * @param delta Indexer Event - not used
+   *
+   * @throws Exception if something goes wrong
+   */ 
+  public void onIndexerEvent(IndexerEvent delta) throws Exception {
+    //Synchronized on the syncObject so we don't miss events
+    synchronized (syncObject) {
+      syncObject.notify();
+    }
+    LOGGER.log(Level.FINEST,"[" + getName() + "] done being notified");
+  }
+
+  /**
+   * Thread main body. Waits until notified, then tries to get the next products and process them.
+   */
+  @Override
+  public void run() {
+    //Run until we're told not to
+    while (!stopThread) {
+
+      List<ProductSummary> productList = null;
+
+      //Synchronized so we aren't notified of new products right before we wait
+      synchronized (syncObject) {
+        try {
+          productList = getNextProducts();
+        } catch (Exception e) {
+          //Not a phenomenal way to handle this exception
+          LOGGER.log(Level.WARNING, "[" + getName() + "] Exception getting next products", e);
```

Comment:
Ideas on how to make this more phenomenal?

Reply:
Updated to handle exception in a similar way to product processing. The behavior (and exceptions) are too abstract for a good solution in this class.